### PR TITLE
fix: validate tuples for non-nil subj

### PR DESCRIPTION
**Pull request by hperl**
This PR adds a validation to reject request where both the subject id and the subject set are nil

### CODE REVIEW

**Review by zepatrik — CHANGES_REQUESTED** [review]
To ensure this regression does not happen again we should add a test :wink:

File: ketoapi/public_api_definitions.go [hunk 1]
```diff
@@ -129,6 +129,13 @@ func (r *RelationTuple) ToLoggerFields() logrus.Fields {
 	return fields
 }
 
+func (r *RelationTuple) Validate() error {
+	if r.SubjectSet == nil && r.SubjectID == nil {
+		return ErrNilSubject
```
Comment (zepatrik) [inline ketoapi/public_api_definitions.go:134]:
```suggestion
		return errors.WithStack(ErrNilSubject)
```

File: internal/relationtuple/transact_server.go [hunk 2]
```diff
@@ -221,6 +226,11 @@ func (h *handler) patchRelationTuples(w http.ResponseWriter, r *http.Request, _
 			h.d.Writer().WriteError(w, r, herodot.ErrBadRequest.WithError("relation_tuple is missing"))
 			return
 		}
+		if err := d.RelationTuple.Validate(); err != nil {
+			h.d.Writer().WriteError(w, r, errors.WithStack(herodot.ErrBadRequest.WithError(err.Error())))
```
Comment (zepatrik) [inline internal/relationtuple/transact_server.go:230]:
```suggestion
			h.d.Writer().WriteError(w, r, err)
```

The error returned by validate should already be a `herodot.ErrBadRequest` with the right details and stack.

File: internal/relationtuple/transact_server.go [hunk 3]
```diff
@@ -111,6 +111,11 @@ func (h *handler) createRelation(w http.ResponseWriter, r *http.Request, _ httpr
 		return
 	}
 
+	if err := rt.Validate(); err != nil {
+		h.d.Writer().WriteError(w, r, errors.WithStack(herodot.ErrBadRequest.WithError(err.Error())))
```
Comment (zepatrik) [inline internal/relationtuple/transact_server.go:115]: same here
```suggestion
		h.d.Writer().WriteError(w, r, err)
```

**Review by zepatrik — COMMENTED** [review]
Just some minor stuff and we're good to go :tada:

File: go.mod [hunk 4]
```diff
@@ -5,6 +5,7 @@ replace (
 	github.com/gogo/protobuf => github.com/gogo/protobuf v1.3.2
 	github.com/mattn/go-sqlite3 => github.com/mattn/go-sqlite3 v1.14.10
 	github.com/oleiade/reflections => github.com/oleiade/reflections v1.0.1
+	github.com/ory/keto/internal/httpclient-next => ./internal/httpclient-next
```
[the remark this answers is not in the document]
Comment (hperl) [reply go.mod:8]: the client has a dependency on "golang.org/x/oauth2". Isn't the whole client autogenerated?

[the remark this answers is not in the document]
Comment (zepatrik) [reply go.mod:8]: Yes, but you could add a `rm internal/httpclient/go.*` to the `make sdk` target.

[the remark this answers is not in the document]
Comment (zepatrik) [reply go.mod:8]: And the dependency will then be tracked in this `go.mod`

[the remark this answers is not in the document]
Comment (hperl) [reply go.mod:8]: agreed, done!

**Comment by hperl** [comment]
@aeneasr this is the PR that contains the removal of the database constraint. I also had to update the generated HTTP client to test for the regression, since the old client never allowed empty strings to be transmitted.

**Review by aeneasr — APPROVED** [review]
